### DISCUSSION TITLE
serverpod_upsert_all module proposal and modularization

### DISCUSSION THREAD

**Discussion by tgrushka** [serverpod/serverpod#1015] [Show and tell]
I'd like to share one of the potential contributions I'm working on, as I needed this feature for an app I'm building. It's called `serverpod_upsert_all` and it's pretty self-explanatory in that it provides an "upsert all" function for Postgres.

Basically it runs a huge insert in batches and only updates rows that have changed, returning the rows that were inserted, updated, or unchanged. It does this with only one SQL statement per batch of default 100 rows.

Still in development, nothing is written in stone, don't have unit tests yet (but testing it extensively as I develop my app), trying to use strong types as much as possible. I had to copy/paste the "serverpod internals" because they are marked private, such as `logQuery` and `formatTableRow`, which I wish could be somehow accessible by modules/plugins.

Would like to offer this as a prototype for how developers might like to extend serverpod with modules and demonstrate some of the needs for integration, and would love to help with this work to modularize serverpod so we can elegantly extend it.

Would appreciate thoughts on whether there is interest in such a module and if publishing it on pub.dev once it can integrate better and has full unit tests might be worthwhile.

https://github.com/dra11y/serverpod_upsert_all

**Comment by tgrushka** [comment]
Here is a very basic code example:

```dart
    final rows = <Book>[
      Book(title: 'To Kill a Mockingbird', author: 'Harper Lee'),
      Book(title: 'Gone with the Wind', author: 'Margaret Mitchell'),
      Book(title: 'Farenheit 451', author: 'Ray Bradbury'),
      Book(title: 'Mio, My Son', author: 'Astrid Lindgren'),
    ];

    await upsertAll(
      session,
      rows: rows,
      onConflict: {Book.t.title, Book.t.author},
    );
```

**Comment by vlidholt** [comment]
There is some initial work done on this feature here:

https://github.com/serverpod/serverpod/pull/580

**Comment by iampopal** [comment]
Looking and need this upsert feature.

**Comment by sedobrengocce** [comment]
#4832 complete the work i left undone in #580

sorry for disappear
